File: csharp/examples/BasicUsage.cs
// Basic usage example for lino-objects-codec C# implementation.

using System;
using System.Collections.Generic;
using Lino.Objects.Codec;

Console.WriteLine("=== lino-objects-codec C# Basic Usage Example ===\n");

// 1. Encode and decode basic types
Console.WriteLine("1. Basic Types:");
Console.WriteLine($"   null: {Codec.Encode(null)}");
Console.WriteLine($"   bool (true): {Codec.Encode(true)}");
Console.WriteLine($"   bool (false): {Codec.Encode(false)}");
Console.WriteLine($"   int: {Codec.Encode(42)}");
Console.WriteLine($"   double: {Codec.Encode(3.14)}");
Console.WriteLine($"   string: {Codec.Encode("Hello, World!")}");
Console.WriteLine();

// 2. Roundtrip a dictionary
Console.WriteLine("2. Dictionary Roundtrip:");
var data = new Dictionary<string, object?>
{
    { "name", "Alice" },
    { "age", 30 },
    { "active", true }
};
var encoded = Codec.Encode(data);
Console.WriteLine($"   Original: {{name: Alice, age: 30, active: true}}");
Console.WriteLine($"   Encoded: {encoded}");

var decoded = Codec.Decode(encoded) as Dictionary<string, object?>;
Console.WriteLine($"   Decoded: {{name: {decoded?["name"]}, age: {decoded?["age"]}, active: {decoded?["active"]}}}");
Console.WriteLine();

// 3. Encode a list
Console.WriteLine("3. List Encoding:");
var list = new List<object?> { 1, 2, 3, "four", true };
encoded = Codec.Encode(list);
Console.WriteLine($"   List: [1, 2, 3, \"four\", true]");
Console.WriteLine($"   Encoded: {encoded}");
Console.WriteLine();

// 4. Nested structures
Console.WriteLine("4. Nested Structure:");
var nested = new Dictionary<string, object?>
{
    {
        "users", new List<object?>
        {
            new Dictionary<string, object?> { { "id", 1 }, { "name", "Alice" } },
            new Dictionary<string, object?> { { "id", 2 }, { "name", "Bob" } }
        }
    },
    {
        "metadata", new Dictionary<string, object?>
        {
            { "version", 1 },
            { "count", 2 }
        }
    }
};
encoded = Codec.Encode(nested);
Console.WriteLine($"   Encoded: {encoded}");
Console.WriteLine();

// 5. Circular references
Console.WriteLine("5. Circular References:");
// Object identity is a property of the compact format, which names shared nodes
// with `obj_N` ids. The readable format writes a plain tree, so it rejects a
// cycle instead of silently unrolling it.

// Self-referencing list
var selfRef = new List<object?>();
selfRef.Add(selfRef);
encoded = Codec.EncodeCompact(selfRef);
Console.WriteLine($"   Self-referencing list encoded: {encoded}");

var decodedSelfRef = Codec.Decode(encoded) as List<object?>;
var isSelfRef = decodedSelfRef != null && ReferenceEquals(decodedSelfRef, decodedSelfRef[0]);
Console.WriteLine($"   Reference preserved after decode: {isSelfRef}");

// Self-referencing dictionary
var selfRefDict = new Dictionary<string, object?>();
selfRefDict["self"] = selfRefDict;
encoded = Codec.EncodeCompact(selfRefDict);
Console.WriteLine($"   Self-referencing dict encoded: {encoded}");
Console.WriteLine();

// 6. Mutual references
Console.WriteLine("6. Mutual References:");
// Also a compact-format property, for the same reason.
var list1 = new List<object?> { 1, 2 };
var list2 = new List<object?> { 3, 4 };
list1.Add(list2);
list2.Add(list1);
encoded = Codec.EncodeCompact(list1);
Console.WriteLine($"   Two lists referencing each other:");
Console.WriteLine($"   {encoded}");

var decodedList1 = Codec.Decode(encoded) as List<object?>;
var decodedList2 = decodedList1?[2] as List<object?>;
var backRef = decodedList2?[2];
Console.WriteLine($"   Circular reference preserved: {ReferenceEquals(decodedList1, backRef)}");
Console.WriteLine();

// 7. Special float values
Console.WriteLine("7. Special Float Values:");
Console.WriteLine($"   NaN: {Codec.Encode(double.NaN)}");
Console.WriteLine($"   Infinity: {Codec.Encode(double.PositiveInfinity)}");
Console.WriteLine($"   -Infinity: {Codec.Encode(double.NegativeInfinity)}");
Console.WriteLine();

// 8. Output formats
Console.WriteLine("8. Output Formats:");
var formats = new Dictionary<string, object?>
{
    { "users", new List<object?> { new Dictionary<string, object?> { { "id", 1 }, { "name", "Alice" } } } },
    { "count", 1 }
};

// The default output is readable: values are written as they are.
var readable = Codec.Encode(formats);
Console.WriteLine("   Readable (default):");
Console.WriteLine(readable);

// The compact form carries a type marker per value and base64 encodes strings.
var compact = Codec.EncodeCompact(formats);
Console.WriteLine($"   Compact: {compact}");

// `Decode` recognises both.
Console.WriteLine($"   Compact document detected: {Codec.IsCompactNotation(compact)}");
Console.WriteLine($"   Readable document detected as compact: {Codec.IsCompactNotation(readable)}");

// A cycle has no readable form; use the compact one.
var cyclic = new Dictionary<string, object?>();
cyclic["self"] = cyclic;
try
{
    Codec.Encode(cyclic);
    Console.Error.WriteLine("   ERROR: a cycle should not be writable as readable text!");
}
catch (CircularReferenceException e)
{
    Console.WriteLine($"   Readable format rejects a cycle: {e.GetType().Name}");
}

Console.WriteLine();

Console.WriteLine("=== Example completed successfully! ===");

File: rust/examples/basic_usage.rs
//! Basic usage example for the lino-objects-codec library.

use lino_objects_codec::{decode, encode, encode_compact, LinoValue};

fn main() {
    println!("=== Links Notation Objects Codec - Rust Example ===\n");

    // Example 1: Basic types
    println!("1. Basic Types:");

    // Null
    let null_val = LinoValue::Null;
    let encoded = encode(&null_val);
    println!(
        "   Null: {} -> decoded: {:?}",
        encoded,
        decode(&encoded).unwrap()
    );

    // Booleans
    let true_val = LinoValue::Bool(true);
    let encoded = encode(&true_val);
    println!(
        "   true: {} -> decoded: {:?}",
        encoded,
        decode(&encoded).unwrap()
    );

    // Integers
    let int_val = LinoValue::Int(42);
    let encoded = encode(&int_val);
    println!(
        "   42: {} -> decoded: {:?}",
        encoded,
        decode(&encoded).unwrap()
    );

    // Floats
    let float_val = LinoValue::Float(3.14159);
    let encoded = encode(&float_val);
    println!(
        "   3.14159: {} -> decoded: {:?}",
        encoded,
        decode(&encoded).unwrap()
    );

    // Special floats
    let inf_val = LinoValue::Float(f64::INFINITY);
    let encoded = encode(&inf_val);
    println!(
        "   Infinity: {} -> decoded: {:?}",
        encoded,
        decode(&encoded).unwrap()
    );

    let nan_val = LinoValue::Float(f64::NAN);
    let encoded = encode(&nan_val);
    let decoded = decode(&encoded).unwrap();
    println!(
        "   NaN: {} -> decoded is_nan: {}",
        encoded,
        decoded.as_float().unwrap().is_nan()
    );

    // Strings
    let str_val = LinoValue::String("Hello, World!".to_string());
    let encoded = encode(&str_val);
    println!(
        "   'Hello, World!': {} -> decoded: {:?}",
        encoded,
        decode(&encoded).unwrap()
    );

    // Unicode strings
    let unicode_val = LinoValue::String("你好世界 🌍".to_string());
    let encoded = encode(&unicode_val);
    println!(
        "   Unicode: {} -> decoded: {:?}",
        encoded,
        decode(&encoded).unwrap()
    );

    println!();

    // Example 2: Collections
    println!("2. Collections:");

    // Array
    let array_val = LinoValue::array([LinoValue::Int(1), LinoValue::Int(2), LinoValue::Int(3)]);
    let encoded = encode(&array_val);
    let decoded = decode(&encoded).unwrap();
    println!("   Array [1, 2, 3]:\n{}", encoded);
    println!("   Decoded: {:?}", decoded);

    // Object
    let obj_val = LinoValue::object([
        ("name", LinoValue::String("Alice".to_string())),
        ("age", LinoValue::Int(30)),
        ("active", LinoValue::Bool(true)),
    ]);
    let encoded = encode(&obj_val);
    let decoded = decode(&encoded).unwrap();
    println!("   Object {{name, age, active}}:\n{}", encoded);
    println!("   Decoded: {:?}", decoded);

    println!();

    // Example 3: Nested structures
    println!("3. Nested Structure:");

    let complex = LinoValue::object([
        ("id", LinoValue::Int(123)),
        ("name", LinoValue::String("Test Object".to_string())),
        (
            "tags",
            LinoValue::array([
                LinoValue::String("tag1".to_string()),
                LinoValue::String("tag2".to_string()),
            ]),
        ),
        (
            "metadata",
            LinoValue::object([
                ("version", LinoValue::Int(1)),
                ("created", LinoValue::String("2025-01-01".to_string())),
            ]),
        ),
    ]);

    let encoded = encode(&complex);
    let decoded = decode(&encoded).unwrap();

    println!("   Encoded:\n{}", encoded);
    println!("   Decoded name: {:?}", decoded.get("name"));
    println!("   Decoded tags: {:?}", decoded.get("tags"));
    println!(
        "   Decoded metadata.version: {:?}",
        decoded.get("metadata").and_then(|m| m.get("version"))
    );

    println!();

    // Example 4: Mixed type array
    println!("4. Mixed Type Array:");

    let mixed = LinoValue::array([
        LinoValue::Int(1),
        LinoValue::String("hello".to_string()),
        LinoValue::Bool(true),
        LinoValue::Null,
        LinoValue::Float(2.5),
    ]);

    let encoded = encode(&mixed);
    let decoded = decode(&encoded).unwrap();

    println!("   Encoded:\n{}", encoded);
    println!("   Decoded: {:?}", decoded);

    println!();

    // Example 5: Roundtrip verification
    println!("5. Roundtrip Verification:");

    let data = LinoValue::object([
        (
            "users",
            LinoValue::array([
                LinoValue::object([
                    ("id", LinoValue::Int(1)),
                    ("name", LinoValue::String("Alice".to_string())),
                ]),
                LinoValue::object([
                    ("id", LinoValue::Int(2)),
                    ("name", LinoValue::String("Bob".to_string())),
                ]),
            ]),
        ),
        ("count", LinoValue::Int(2)),
    ]);

    let encoded = encode(&data);
    let decoded = decode(&encoded).unwrap();

    println!("   Original == Decoded: {}", data == decoded);

    println!();

    // Example 6: The compact (base64) form is still available under its own name
    println!("6. Compact Form:");

    let compact = encode_compact(&data);
    println!("   Compact: {}", compact);
    println!(
        "   Compact decodes back to the same value: {}",
        decode(&compact).unwrap() == data
    );

    println!("\n=== Example completed successfully! ===");
}
